Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIPermissions.mm

```diff
@@ -0,0 +1,204 @@
+
+    auto configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
+    configuration.get()._webExtensionController = manager.get().controller;
+
+    [manager loadAndRun];
+}
+
+TEST(WKWebExtensionAPIPermissions, AcceptPermissionsRequestTest)
+{
+    auto *manifest = @{
+        @"manifest_version": @3,
+        @"permissions": @[ @"alarms", @"activeTab" ],
+        @"optional_permissions": @[ @"webNavigation", @"cookies", @"declarativeNetRequest", @"*://*.apple.com/*" ],
+        @"background": @{ @"scripts": @[ @"background.js" ], @"type": @"module", @"persistent": @NO },
+        @"host_permissions": @[ @"*://*.apple.com/*" ]
+    };
+
+    auto *backgroundScript = Util::constructScript(@[
+        // Finish
+        @"browser.test.yield()",
+        @"browser.test.notifyPass()"
+    ]);
+
+    auto extension = adoptNS([[_WKWebExtension alloc] _initWithManifestDictionary:manifest resources:@{ @"background.js": backgroundScript }]);
+    auto manager = adoptNS([[TestWebExtensionManager alloc] initForExtension:extension.get()]);
+
+    auto configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
+    configuration.get()._webExtensionController = manager.get().controller;
+
+    NSSet<NSString *> *permissions = [NSSet setWithArray:@[ @"declarativeNetRequest" ]];
+    NSSet<_WKWebExtensionMatchPattern *> *matchPatterns = [NSSet setWithArray:@[ [_WKWebExtensionMatchPattern matchPatternWithString:@"*://*.apple.com/*" ]]];
+    auto requestDelegate = adoptNS([[TestWebExtensionsDelegate alloc] init]);
+    __block bool requestComplete = false;
+
+    // Implement the delegate methods that're called when a call to permissions.request() is made.
+    requestDelegate.get().promptForPermissions = ^(id<_WKWebExtensionTab> tab, NSSet<NSString *> *requestedPermissions, void (^callback)(NSSet<NSString *> *)) {
+        EXPECT_EQ(requestedPermissions.count, permissions.count);
+        EXPECT_TRUE([requestedPermissions isEqualToSet:permissions]);
+        callback(requestedPermissions);
+    };
+
+    requestDelegate.get().promptForPermissionMatchPatterns = ^(id<_WKWebExtensionTab> tab, NSSet<_WKWebExtensionMatchPattern *> *requestedMatchPatterns, void (^callback)(NSSet<_WKWebExtensionMatchPattern *> *)) {
+        EXPECT_EQ(requestedMatchPatterns.count, matchPatterns.count);
+        EXPECT_TRUE([requestedMatchPatterns isEqualToSet:matchPatterns]);
+        callback(requestedMatchPatterns);
+        requestComplete = true;
+    };
+
+    manager.get().controllerDelegate = requestDelegate.get();
+    [manager loadAndRun];
+
+    runScriptWithUserGesture("browser.permissions.request({'permissions': ['declarativeNetRequest'], 'origins': ['*://*.apple.com/*']})"_s, manager.get().context._backgroundWebView);
+
+    TestWebKitAPI::Util::run(&requestComplete);
+}
+
+TEST(WKWebExtensionAPIPermissions, DenyPermissionsRequestTest)
```

Comment:
I'm struggling to figure out how to get this test to pass. I confirmed that the call to request returns `false` by adding logs, but for some reason, the test hangs and timeout without a useful error message. Any suggestions on how I can restructure this test?

Reply:
Shouldn't the call to `request()` return `true` since you grant what was requested in the blocks? If it returns `false` that suggests it is returning easy somewhere and/or `requestComplete = true` is not being reached.

The structure of the test looks correct to me.

Reply:
Also can you add `browser.test.assertTrue` around `browser.permissions.request()` like you do for the deny case below?

Reply:
Spoke to Tim about this offline. The DenyRequestTest should return false since it's not granting permissions. I added `browser.test.assertTrue` to the AcceptRequestTest. Was able to get the second test to pass.